### ISSUE TITLE
Treatment of 0 values as missing in xgboost predict (R)

### ISSUE THREAD

**Issue by gilad19** [dmlc/xgboost#1842]
For bugs or installation issues, please provide the following information.
The more information you provide, the more easily we will be able to offer
help and advice.

## Environment info
Operating System:
Linux

Compiler:

Package used (python/R/jvm/C++):
XGBoost

`xgboost` version used:
0.4-4

If installing from source, please provide 

1. The commit hash (`git rev-parse HEAD`)
2. Logs will be helpful (If logs are large, please upload as attachment).

If you are using python package, please provide

1. The python version and distribution
2. The command to install `xgboost` if you are not installing from source

If you are using R package, please provide

1. The R `sessionInfo()`
R version 3.3.0 (2016-05-03)
Platform: x86_64-redhat-linux-gnu (64-bit)
Running under: Oracle Linux Server 7.2

locale:
 [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C               LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    
 [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8    LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] DiagrammeR_0.8.4 r2pmml_0.8.1     xgboost_0.4-4    data.table_1.9.6

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.5      lattice_0.20-33  visNetwork_1.0.2 digest_0.6.9     plyr_1.8.4       chron_2.3-47     grid_3.3.0      
 [8] jsonlite_1.0     magrittr_1.5     scales_0.4.0     stringi_1.1.1    SparseM_1.7      rstudioapi_0.6   Matrix_1.2-6    
[15] tools_3.3.0      stringr_1.0.0    htmlwidgets_0.6  igraph_1.0.1     munsell_0.4.3    yaml_2.1.13      influenceR_0.1.0
[22] colorspace_1.2-6 htmltools_0.3.5 

2. The command to install `xgboost` if you are not installing from source
install.packages("xgboost")

## 0 values
3. It seems that when prediction is used on regular matrix (not sparse) which includes 0 values (which represent real values and not missing values) they are treated as missing values and the wrong split is chosen (e.g. . when a sample had 0 value it was split to the missing node of the tree and not the correct one). 

I tried setting missing = NA or missing = -555555555555 during the training but it didn't change the outcome.

Is there a way to not treat 0 values as missing values during prediction?

Here is the code I used:

xgb_model <- xgboost(
  data = trainMatrix,
  label = ifelse(trainMatrix[,1] == "CLEAN", 0, 1),
  nthread = 8,
  metrics = "auc",
  objective = "binary:logistic",
  verbose = TRUE,
  nround = 5,
  missing = NA
)

predict(xgb_model, testMatrix)

**Comment by khotilov** [comment]
The xgboost code that you are using is quite old. And when you set a specific `missing=` value for training data, the same setting should be used for testing data as well.

In the new code, things should be fairly consistent: 
- `missing = NA` by default
- sparse elements in sparse data are also treated as "missing" by the gbtree booster (thus, don't mix sparse and dense data in training and testing/prediction)
- the gblinear booster treats missing values as zeros

**Comment by gilad19** [comment]
Thanks.

I downloaded the most updated xgboost code I found (0.6-0?) and indeed notice that there the default for missing is NA (and not NULL as in the CRAN version). However, I still see the same behavior, 0 is treated as missing value when the tree is doing the split. 

I also tried to define implicitly "missing = NA" for training/testing/prediction with the old and new code and got the same results. 

By the way, when I use the same data in Java I get the correct results and 0 is not treated as missing value. 
Any help will be highly appreciated.

**Comment by khotilov** [comment]
It would be helpful if you could illustrate your observations with an example.

Note that for some kinds of data, e.g., data similar to binary hot-encoding in dense format, all missing branch directions might coincide with the directions of zeros - it just might be that such missing directions were actually learned.
